File: php/infer_model.php
<?php
// php/infer_model.php — clearer inference adapter for the trained pose classifier
require_once 'config.php';
header('Content-Type: application/json');
startSecureSession();

function json_error(string $msg, int $code = 400) {
    http_response_code($code);
    echo json_encode(['error' => $msg]);
    exit;
}

if (empty($_SESSION['user_id'])) {
    json_error('Unauthorized', 401);
}

$payload = json_decode(file_get_contents('php://input'), true) ?? [];
$featureVector = $payload['feature_vector'] ?? null;
if ($featureVector === null || !is_array($featureVector)) {
    json_error('Missing or invalid "feature_vector" payload');
}

$modelPath = __DIR__ . '/../python/model/pose_classifier.json';
$scalerPath = __DIR__ . '/../python/model/pose_scaler.json';
if (!file_exists($modelPath) || !file_exists($scalerPath)) {
    json_error('Model files not found. Run python/train_model.py first.', 500);
}

$model = json_decode(file_get_contents($modelPath), true);
$scaler = json_decode(file_get_contents($scalerPath), true);
if (!is_array($model) || !is_array($scaler)) {
    json_error('Failed to load model files', 500);
}

$featureNames = $scaler['feature_names'] ?? [];
if (!is_array($featureNames) || count($featureNames) === 0) {
    json_error('Scaler missing feature names', 500);
}

// Build ordered input vector. Accept either a name=>value map or a numeric array
$inputVector = [];
foreach ($featureNames as $feature) {
    if (array_key_exists($feature, $featureVector)) {
        $inputVector[] = floatval($featureVector[$feature]);
    } else {
        // fallback: if payload sent numeric array in same order
        $inputVector[] = isset($featureVector[count($inputVector)]) ? floatval($featureVector[count($inputVector)]) : 0.0;
    }
}

// Standardize
$mean = $scaler['mean'] ?? [];
$scale = $scaler['scale'] ?? [];
if (!is_array($mean) || !is_array($scale)) json_error('Invalid scaler metadata', 500);
$standardized = [];
for ($i = 0; $i < count($inputVector); $i++) {
    $m = $mean[$i] ?? 0.0;
    $s = $scale[$i] ?? 1.0;
    $standardized[] = ($s != 0.0) ? (($inputVector[$i] - $m) / $s) : 0.0;
}

// Helpers
function relu(array $x): array { return array_map(fn($v) => max(0.0, $v), $x); }
function softmax(array $x): array { $mx = max($x); $ex = array_map(fn($v) => exp($v - $mx), $x); $sum = array_sum($ex); return array_map(fn($v) => $v / ($sum ?: 1.0), $ex); }

// Run through the MLP model weights saved by train_model.py
$activations = $standardized;
foreach ($model['coefs'] as $layerIndex => $coefMatrix) {
    $next = [];
    $intercepts = $model['intercepts'][$layerIndex];
    foreach ($intercepts as $j => $bias) {
        $sum = $bias;
        foreach ($activations as $i => $val) {
            $w = $coefMatrix[$i][$j] ?? 0.0;
            $sum += $val * $w;
        }
        $next[] = $sum;
    }
    $activations = ($layerIndex < count($model['coefs']) - 1) ? relu($next) : $next;
}

$probs = softmax($activations);
$bestIdx = array_search(max($probs), $probs, true);
$classNames = $model['class_names'] ?? $model['classes'] ?? [];
$prediction = $classNames[$bestIdx] ?? null;

echo json_encode([
    'exercise' => $prediction,
    'confidence' => round(($probs[$bestIdx] ?? 0.0) * 100, 1),
    'probabilities' => $probs
]);
exit;

File: python/infer_model.py
import json
from pathlib import Path

import numpy as np

MODEL_DIR = Path(__file__).resolve().parent / 'model'
MODEL_FILE = MODEL_DIR / 'pose_classifier.json'
SCALER_FILE = MODEL_DIR / 'pose_scaler.json'


def load_json(path):
    with open(path, 'r', encoding='utf-8') as f:
        return json.load(f)


def softmax(x):
    exp = np.exp(x - np.max(x))
    return exp / exp.sum(axis=-1, keepdims=True)


class PoseClassifier:
    def __init__(self):
        self.model = load_json(MODEL_FILE)
        self.scaler = load_json(SCALER_FILE)
        self.feature_names = self.scaler['feature_names']
        self.mean = np.array(self.scaler['mean'], dtype=np.float32)
        self.scale = np.array(self.scaler['scale'], dtype=np.float32)

    def transform(self, X):
        X = np.asarray(X, dtype=np.float32)
        return (X - self.mean) / self.scale

    def predict(self, X):
        X = self.transform(X)
        activations = X
        for i, (coef, intercept) in enumerate(zip(self.model['coefs'], self.model['intercepts'])):
            activations = activations.dot(np.array(coef, dtype=np.float32)) + np.array(intercept, dtype=np.float32)
            if i < len(self.model['coefs']) - 1:
                activations = np.maximum(activations, 0)
        probs = softmax(activations)
        probs = np.array(probs)
        # handle both (n_classes,) and (batch, n_classes) shapes
        if probs.ndim == 2:
            label_index = int(np.argmax(probs, axis=1)[0])
        else:
            label_index = int(np.argmax(probs))
        return label_index, probs.tolist()


if __name__ == '__main__':
    clf = PoseClassifier()
    sample = np.zeros(len(clf.feature_names), dtype=np.float32)
    label, probs = clf.predict(sample.reshape(1, -1))
    print('predicted', label, probs)
